src/base/init/init.c

```diff
@@ -405,15 +403,31 @@ void low_mem_init(void)
   }
 
   if (config.ext_mem) {
-    /* LOWMEM_SIZE + HMASIZE == base
-     * Use dpmi_rsv_low as ext_mem. */
-    register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE,
-	    MEM_BASE32(LOWMEM_SIZE + HMASIZE), LOWMEM_SIZE + HMASIZE);
+    /* LOWMEM_SIZE + HMASIZE == base */
     memcheck_addtype('X', "EXT MEM");
     memcheck_reserve('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE);
     x_printf("Ext.Mem of size 0x%x at %#x\n", EXTMEM_SIZE, result + LOWMEM_SIZE + HMASIZE);
   }
 
+  /* create non-identity mapping up to dpmi_base */
+  phys_rsv = config.dpmi_base - (LOWMEM_SIZE + HMASIZE);
+  ptr = smalloc_aligned_topdown(&main_pool, MEM_BASE32(memsize),
+	PAGE_SIZE, phys_rsv);
+  assert(ptr);
+  /* establish alias access for int15 */
+  register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
+	    lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr));
+  register_hardware_ram_virtual('L',  DOSADDR_REL(ptr), phys_rsv,
+	    MEM_BASE32(LOWMEM_SIZE + HMASIZE), LOWMEM_SIZE + HMASIZE);
+  /* and identity map */
+  register_hardware_ram_virtual('D',  config.dpmi_base, dpmi_mem_size(),
```

Comment:
this breaks `test/test_dos.py PPDOSGITTestCase.test_memory_emm286_borland` again, as `dpmi_mem_size` doesn't include the high linear DPMI address space. You could fix it like in my patch (but without the -1 for now):
```
diff --git a/src/base/init/init.c b/src/base/init/init.c
index 1ae54a710..7096f7de3 100644
--- a/src/base/init/init.c
+++ b/src/base/init/init.c
@@ -420,7 +420,8 @@ void low_mem_init(void)
   register_hardware_ram_virtual('L',  DOSADDR_REL(ptr), phys_rsv,
            MEM_BASE32(LOWMEM_SIZE + HMASIZE), LOWMEM_SIZE + HMASIZE);
   /* and identity map */
-  register_hardware_ram_virtual('D',  config.dpmi_base, dpmi_mem_size(),
+  register_hardware_ram_virtual('D',  config.dpmi_base,
+           DOSADDR_REL(ptr) - (LOWMEM_SIZE + HMASIZE),
            MEM_BASE32(config.dpmi_base), config.dpmi_base);
   /* create alias for dpmi */
   result = alias_mapping(MAPPING_EXTMEM, DOSADDR_REL(ptr), EXTMEM_SIZE,
```

Reply:
`+           DOSADDR_REL(ptr) - (LOWMEM_SIZE + HMASIZE),`

But this is likely wrong, did you mean
`+           DOSADDR_REL(ptr) - config.dpmi_base,`?
But I decided to special-case that, as
creating physical region for uncommitted
mapping sounds weird.

Reply:
yes you are correct. I forgot I merged the two in my patch

Reply:
Yes, done.